# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/Bailouts.cpp
# Commit: 7f75f8b705e7
# Full Hash: 7f75f8b705e72090c3faa76aa8bac440c6d0fde6
# Author: Iain Ireland <iireland@mozilla.com>
# Date: 2022-03-31 09:35:41
# Regressor Bug: 885514
# File Overlap Count: 2
# Description:
#   Bug 885514: Support bailouts for finally r=jandem
#   
#   This is mostly modeled after the catch code. The main difference is that finally expects to resume in code where there are two additional values on the stack, so we store the exception in the ExceptionBailoutInfo and write it into the stack frame in BaselineStackBuilder.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D142118
# ==============================================================================

diff -r 8cd3f1a267ae -r 7f75f8b705e7 js/src/jit/Bailouts.cpp
--- a/js/src/jit/Bailouts.cpp	Wed Mar 30 23:19:47 2022 +0000
+++ b/js/src/jit/Bailouts.cpp	Wed Mar 30 23:19:47 2022 +0000
@@ -203,11 +203,14 @@
                                   const InlineFrameIterator& frame,
                                   ResumeFromException* rfe,
                                   const ExceptionBailoutInfo& excInfo) {
-  // We can be propagating debug mode exceptions without there being an
+  // If we are resuming in a finally block, the exception has already
+  // been captured.
+  // We can also be propagating debug mode exceptions without there being an
   // actual exception pending. For instance, when we return false from an
   // operation callback like a timeout handler.
-  MOZ_ASSERT_IF(!excInfo.propagatingIonExceptionForDebugMode(),
-                cx->isExceptionPending());
+  MOZ_ASSERT_IF(
+      !cx->isExceptionPending(),
+      excInfo.isFinally() || excInfo.propagatingIonExceptionForDebugMode());
 
   JS::AutoSaveExceptionState savedExc(cx);
 
@@ -235,6 +238,8 @@
     if (excInfo.propagatingIonExceptionForDebugMode()) {
       bailoutInfo->bailoutKind =
           mozilla::Some(BailoutKind::IonExceptionDebugMode);
+    } else if (excInfo.isFinally()) {
+      bailoutInfo->bailoutKind = mozilla::Some(BailoutKind::Finally);
     }
 
     rfe->kind = ResumeFromException::RESUME_BAILOUT;